internal/clients/redis/producer.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package redis
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/go-redis/redis/v8"
+)
+
+const (
+	checkUnpublishedEventsInterval = 1 * time.Minute
+	checkRedisConnectionInterval   = 100 * time.Millisecond
+)
+
+// Event represents redis event.
+type Event interface {
+	// Encode encodes event to map.
+	Encode() (map[string]interface{}, error)
+}
+
+// Publisher specifies redis event publishing API.
+type Publisher interface {
+	// Publishes event to redis stream.
+	Publish(ctx context.Context, event Event) error
+
+	// StartPublishingRoutine starts routine that checks for unpublished events
+	// and publishes them to redis stream.
+	StartPublishingRoutine(ctx context.Context)
+}
+
+type eventStore struct {
+	client            *redis.Client
+	unpublishedEvents []*redis.XAddArgs
```

Comment:
this can grow unexpectedly large sizes. I think a different approach is required for memory management. Perhaps a channel or some external queue like nats might serve better

Reply:
Currently, nats is configured as fire and forget unless we move to Jetstream. I have implemented a channel with a max of 1M capacity.

Reply:
Before publishing the memory footprint of the container was 17MB after publishing before resuming Redis it was 49MB